spatialmath/capsule.go

```diff
@@ -0,0 +1,346 @@
+	return c.segA.Sub(s.pose.Point()).Norm()+c.radius <= s.radius && c.segB.Sub(s.pose.Point()).Norm()+c.radius <= s.radius
+}
+
+// capsuleBoxSeparatingAxis returns immediately as soon as any result is found indicating that the two objects are not in collision.
+func capsuleBoxSeparatingAxis(a *capsule, b *box) bool {
+	capCenter := a.segA.Add(a.segB).Mul(0.5)
+	centerDist := b.pose.Point().Sub(capCenter)
+
+	// check if there is a distance between bounding spheres to potentially exit early
+	if centerDist.Norm()-((a.length/2)+b.boundingSphereR) > CollisionBuffer {
+		return false
+	}
+	rmA := a.pose.Orientation().RotationMatrix()
+	rmB := b.pose.Orientation().RotationMatrix()
+
+	segLen := (a.length - 2*a.radius) / 2
+	capVec := rmA.Row(2).Mul(segLen)
+
+	for i := 0; i < 3; i++ {
+		if separatingAxisTest1D(centerDist, rmA.Row(i), capVec, b.halfSize, rmB) > CollisionBuffer+a.radius {
+			return false
+		}
+		if separatingAxisTest1D(centerDist, rmB.Row(i), capVec, b.halfSize, rmB) > CollisionBuffer+a.radius {
+			return false
+		}
+		for j := 0; j < 3; j++ {
+			crossProductPlane := rmA.Row(i).Cross(rmB.Row(j))
+
+			// if edges are parallel, this check is already accounted for by one of the face projections, so skip this case
+			if !utils.Float64AlmostEqual(crossProductPlane.Norm(), 0, floatEpsilon) {
+				if separatingAxisTest1D(centerDist, crossProductPlane, capVec, b.halfSize, rmB) > CollisionBuffer+a.radius {
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
+
+func capsuleBoxSeparatingAxisDistance(a *capsule, b *box) float64 {
+	capCenter := a.segA.Add(a.segB).Mul(0.5)
+	centerDist := b.pose.Point().Sub(capCenter)
+
+	// check if there is a distance between bounding spheres to potentially exit early
+	if boundingSphereDist := centerDist.Norm() - ((a.length / 2) + b.boundingSphereR); boundingSphereDist > CollisionBuffer {
+		return boundingSphereDist
+	}
+	rmA := a.pose.Orientation().RotationMatrix()
+	rmB := b.pose.Orientation().RotationMatrix()
+
+	segLen := (a.length - 2*a.radius) / 2
+	capVec := rmA.Row(2).Mul(segLen)
+
+	max := math.Inf(-1)
+	for i := 0; i < 3; i++ {
+		if separation := separatingAxisTest1D(centerDist, rmA.Row(i), capVec, b.halfSize, rmB); separation > max {
+			max = separation
+		}
+		if separation := separatingAxisTest1D(centerDist, rmB.Row(i), capVec, b.halfSize, rmB); separation > max {
+			max = separation
+		}
+		for j := 0; j < 3; j++ {
+			crossProductPlane := rmA.Row(i).Cross(rmB.Row(j))
+
+			// if edges are parallel, this check is already accounted for by one of the face projections, so skip this case
+			if !utils.Float64AlmostEqual(crossProductPlane.Norm(), 0, floatEpsilon) {
+				if separation := separatingAxisTest1D(centerDist, crossProductPlane, capVec, b.halfSize, rmB); separation > max {
+					max = separation
+				}
+			}
+		}
+	}
+	return max - a.radius
+}
+
+func separatingAxisTest1D(positionDelta, plane, capVec r3.Vector, halfSizeB [3]float64, rmB *RotationMatrix) float64 {
```

Comment:
is this not duplicating code?

Reply:
Not quite; because the capsule is being modeled here as a 0x0xN box, we can only do a single `	sum -= math.Abs(capVec.Dot(plane))` rather than having to iterate through two half-sizes that are known to be zero.

A box-capsule SAT check only takes ~40ns on my computer. Adding 4 nanoseconds of computation here would mean collision checks run 10% slower. So in this case it's worth it to duplicate for performance.

Even things like passing in function pointers to specify which separating axis test function to use to reduce duplication, would be slower, because it will mess with compiler optimizations and also with branch prediction inside the CPU itself. Currently all of these loops are able to be fully unrolled and very accurately branch-predicted, which is *fast*